### PR TITLE
[PyTorch] Add CUDA graph tests with FP8 weight caching

### PR DESCRIPTION
**Pull request by timmoon10**
# Description

Our CUDA graph infrastructure (https://github.com/NVIDIA/TransformerEngine/pull/575) supports FP8 weight caching when training with multiple gradient accumulation steps. While adding tests for this functionality (see https://github.com/NVIDIA/TransformerEngine/pull/820#issuecomment-2127600173), I ran into subtle correctness issues because `te.make_graphed_callables` resets grad buffers after capturing graphs, so the gradient buffer filled in the backward pass is different from the gradient buffer used by the optimizer. Note that we didn't detect this before because Megatron-LM and Nemo explicitly manage gradient buffers, e.g. with a distributed optimizer.

This PR modifies the CUDA graph tests to initialize grads before `make_graphed_callables` and it avoids resetting grads within `make_graphed_callables`.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Changes

Please list the changes introduced in this PR:

- Add CUDA graph tests for FP8 weight caching
- Do not reset parameter gradients after graph capture

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by timmoon10** [comment]
Note that users must also initialize grads before calling [`torch.cuda.make_graphed_callables`](https://pytorch.org/docs/stable/generated/torch.cuda.make_graphed_callables.html), or else they'll run into similar correctness issues as I ran into. Consider the following example:
```python
import torch
torch.set_default_device('cuda')

# Construct linear module
model = torch.nn.Linear(1, 1, bias=False)
with torch.no_grad():
    model.weight.fill_(1)
    # model.weight.grad = torch.empty_like(model.weight)  # Uncomment to fix bug

# Capture CUDA graph
x = torch.ones((1, 1), requires_grad=True)
model = torch.cuda.make_graphed_callables(model, (x,))

# Training steps
for step in range(3):
    if model.weight.grad is not None:
        model.weight.grad.zero_()
    x = torch.ones((1, 1), requires_grad=True)
    y = model(x)
    y.backward(torch.ones((1, 1)))
    print(f"{step=}, {model.weight.grad.item()=}")
```
I expect the weight gradient to always be 1. However:
```
step=0, model.weight.grad.item()=1.0
step=1, model.weight.grad.item()=2.0
step=2, model.weight.grad.item()=2.0
```

File: tests/pytorch/test_cuda_graphs.py [hunk 1]
```diff
@@ -199,7 +201,7 @@ def _test_cuda_graphs(
 
     # Loss function and optimizer.
     if not dpa:
-        optimizer = optimizer(model.parameters(), lr=0.001)
+        optimizer = torch.optim.SGD(model.parameters(), lr=0.001)
```
Comment (nvcforster) [inline tests/pytorch/test_cuda_graphs.py:230]: In dropping the adam optimizer from the test parameterization, are we losing any test coverage w.r.t. cuda graphs?

**Review by nvcforster — COMMENTED** [review]
I left a question about coverage, otherwise LGTM.

File: tests/pytorch/test_cuda_graphs.py (hunk 1 above)
Comment (timmoon10) [reply tests/pytorch/test_cuda_graphs.py:230]: This test doesn't really test convergence, but just checks that results match exactly with and without CUDA graphs. SGD and Adam are implemented similarly, so we just need to test one to make sure that the CUDA graph infrastructure is working correctly.

File: tests/pytorch/test_cuda_graphs.py [hunk 2]
```diff
@@ -114,101 +158,134 @@ def _test_cuda_graphs(config, bs, num_layers, dtype, fp8, fp8_params, graph, mod
...
 
     # Launch.
-    for _ in range(10):
-        inputs, target = generate_data(config.s, bs, config.h, config.nheads, config.kv, dtype, dpa=dpa, gen_labels=True)
-        with fp8_autocast(enabled=fp8):
-            output = model(*inputs)
-        loss = loss_fn(output, target)
-        loss.backward()
+    for train_step in range(3):
```
[the remark this answers is not in the document]
Comment (timmoon10) [reply tests/pytorch/test_cuda_graphs.py:233]: Good catch, it was left over from debugging.
```suggestion
    for _ in range(3):
```
